src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -229,10 +232,11 @@ def finalizer():
           stderr.flush()
         finally:
           time.sleep(.001)  # HACK: Sleep 1ms in the main thread to free the GIL.
-          writer.stop()
+          stdout_pipe.stop_writing()
+          stderr_pipe.stop_writing()
           writer.join()
-          stdout.close()
-          stderr.close()
+          if writer.isAlive():
+            raise NailgunStreamWriterError("pantsd timed out while waiting for the stdout/err to finish writing to the socket.")
```

Comment:
It might be useful to join with a (long: 60s?) timeout to encourage this error in unusual cases?